File: modules/chinet/src/Session.cpp
#include "Session.h"
#include "Port.h"


bool Session::read_from_db(const std::string &oid_string){
    bool return_value = true;
    return_value &= MongoObject::read_from_db(oid_string);
    return_value &= create_and_connect_objects_from_oid_array(&document, "nodes", &nodes);
    return return_value;
}

bool Session::write_to_db(){

    bool re = MongoObject::write_to_db();

    for(auto &o : nodes){
        if(!o.second->is_connected_to_db()){
            re &= connect_object_to_db(o.second);
        }
        o.second->write_to_db();
    }

    return re;
}

bson_t Session::get_bson(){
    bson_t doc = MongoObject::get_bson_excluding("nodes", NULL);
    create_oid_array_in_doc(&doc, "nodes", nodes);
    return doc;
}

std::shared_ptr<Port> Session::create_port(
        json port_template,
        std::string port_key
    ) {
#if CHINET_VERBOSE
    std::clog << "CREATE PORT" << std::endl;
#endif
    auto port = std::make_shared<Port>();
    port->set_name(port_key);
#if CHINET_VERBOSE
    std::clog << "-- port name: " << port_key << std::endl;
#endif
    for (json::iterator it_val = port_template.begin(); it_val != port_template.end(); ++it_val) {
        if (it_val.key() == "is_fixed") {
            bool is_fixed = port_template["is_fixed"].get<bool>();
            port->set_fixed(is_fixed);
#if CHINET_VERBOSE
            std::clog << "-- is_fixed: " << is_fixed << std::endl;
#endif
        } else if (it_val.key() == "value") {
            auto b = port_template["value"].get<std::vector<double>>();
#if CHINET_VERBOSE
            std::clog << "-- number of values: " << b.size() << std::endl;
#endif
            bool is_fixed = port->is_fixed();
            port->set_fixed(false);
            port->set_value(b.data(), b.size());
            port->set_fixed(is_fixed);
        } else if (it_val.key() == "is_output"){
            bool is_output = port_template["is_output"].get<bool>();
#if CHINET_VERBOSE
            std::clog << "-- is_output: " << is_output << std::endl;
#endif
            port->set_port_type(is_output);
        } else if (it_val.key() == "is_reactive"){
            bool is_reactive = port_template["is_reactive"].get<bool>();
#if CHINET_VERBOSE
            std::clog << "-- is_reactive: " << is_reactive << std::endl;
#endif
            port->set_reactive(is_reactive);
        }
    }
    return port;
}

std::shared_ptr<Port> Session::create_port(char* port_template, char* port_key){
#if CHINET_VERBOSE
    std::clog << "Session:" << port_template << ":" << port_template << std::endl;
#endif
    return create_port(json::parse(port_template), port_key);
}

std::shared_ptr<Node> Session::create_node(json node_template, std::string node_key){
#if CHINET_VERBOSE
    std::clog << "CREATE NODE" << std::endl;
#endif
    auto node = std::make_shared<Node>(node_key);
    std::string callback;
    std::string callback_type;
    for (json::iterator it = node_template.begin(); it != node_template.end(); ++it) {
        if (it.key() == "ports") {
#if CHINET_VERBOSE
            std::clog << "-- adding ports... " << std::endl;
#endif
            auto ports_json = node_template["ports"];
            for (json::iterator it2 = ports_json.begin(); it2 != ports_json.end(); ++it2) {
                const std::string &port_key = it2.key();
#if CHINET_VERBOSE
                std::clog << "-- adding port key: " << port_key << std::endl;
#endif
                auto port_json = node_template["ports"][port_key];
                auto port = create_port(port_json, port_key);
                node->add_port(port_key, port, port->is_output());
            }
        } else if (it.key() == "callback") {
            callback = node_template["callback"].get<std::string>();
#if CHINET_VERBOSE
            std::clog << "-- callback: " << callback << std::endl;
#endif
        } else if (it.key() == "callback_type") {
            callback_type = node_template["callback_type"].get<std::string>();
#if CHINET_VERBOSE
            std::clog << "-- callback_type: " << callback_type << std::endl;
#endif
        }
    }
    node->set_callback(callback, callback_type);
    return node;
}

std::shared_ptr<Node> Session::create_node(char* node_template, char* port_key){
    return create_node(
            json::parse(node_template),
            port_key
    );
}

bool Session::read_session_template(const std::string &json_string){
#if CHINET_VERBOSE
    std::clog << "READ SESSION TEMPLATE" << std::endl;
#endif
    json session_json = json::parse(json_string);
    // read nodes
    json nodes_json = session_json["nodes"];
    for (json::iterator it = nodes_json.begin(); it != nodes_json.end(); ++it) {
        auto node_key = it.key();
        add_node(node_key, create_node(nodes_json[node_key], node_key));
    }
    auto l = session_json["links"];
    for (json::iterator it = l.begin(); it != l.end(); ++it) {
        auto v = it.value();
        link_nodes(
                v["node"].get<std::string>(),
                v["port"].get<std::string>(),
                v["target_node"].get<std::string>(),
                v["target_port"].get<std::string>()
                );
    }
    return true;
}

bool Session::link_nodes(
        const std::string &node_name,
        const std::string &port_name,
        const std::string &target_node_name,
        const std::string &target_port_name){
    auto itn = nodes.find(node_name);
    auto itnt = nodes.find(target_node_name);

    if(itn != nodes.end() && itnt != nodes.end()){
        auto ports = nodes[node_name]->get_ports();
        auto target_ports = nodes[target_node_name]->get_ports();
        auto itp = ports.find(port_name);
        auto itpt = ports.find(target_port_name);
        if(itp != ports.end() && itpt != target_ports.end()){
            ports[port_name]->set_link(target_ports[target_port_name]);
            return true;
        }
    }
    return false;
}

void Session::add_node(
        std::string name,
        std::shared_ptr<Node> object
)
{
    nodes[name] = object;
    object->set_name(name);
    if (is_connected_to_db()) {
        connect_object_to_db(object);
    }
}

std::map<std::string, std::shared_ptr<Node>> Session::get_nodes()
{
    return nodes;
}

std::string Session::get_session_template(){
    return "";
}

File: modules/chinet/chinet/session.py
import json
from .db import DB
from .base import BaseObject
from .node import Node
from .port import Port

class Session(BaseObject):
    """
    Enhanced Session class for managing a graph of Nodes and Ports.
    Mirrors the C++ Session class logic 1:1.
    """
    def __init__(self, nodes=None):
        super().__init__(name="session")
        self.nodes = {}
        if nodes:
            for name, node in nodes.items():
                self.add_node(name, node)
        self._document["type"] = "session"
        DB.register(self)

    def add_node(self, name, node):
        self.nodes[name] = node
        DB.register(node)
        for p in node.get_ports().values():
            DB.register(p)

    def get_nodes(self):
        return self.nodes

    def write_to_db(self):
        res = super().write_to_db()
        for node in self.nodes.values():
            if not node.is_connected_to_db:
                self.connect_object_to_db(node)
            res &= node.write_to_db()
        return res

    def read_from_db(self, oid):
        self.oid = oid
        # Try to recover state from global DB if already present
        stored = DB.get(oid)
        if stored:
            if isinstance(stored, Session):
                self.nodes = stored.nodes.copy()
                self.set_document(stored._document)
                return True
            elif isinstance(stored, dict):
                # It's a document from DB registry
                self.set_document(stored)
                node_oids = self._document.get("nodes", {})
                self.nodes.clear()
                for name, n_oid in node_oids.items():
                    node = DB.get(n_oid)
                    if node: self.nodes[name] = node
                return True
        
        # Fallback to base read_from_db
        res = super().read_from_db(oid)
        if res:
            node_oids = self._document.get("nodes", {})
            self.nodes.clear()
            for name, n_oid in node_oids.items():
                node = DB.get(n_oid)
                if node: self.nodes[name] = node
            return True
        return False

    def create_port(self, port_template, port_key):
        if isinstance(port_template, str): port_template = json.loads(port_template)
        p = Port(name=port_key, **port_template)
        if "value" in port_template: p.value = port_template["value"]
        DB.register(p)
        return p

    def create_node(self, node_template, node_key):
        if isinstance(node_template, str): node_template = json.loads(node_template)
        n = Node(name=node_key)
        n.set_callback(node_template.get("callback", ""), node_template.get("callback_type", ""))
        self.add_node(node_key, n)
        return n

    def read_session_template(self, json_string): return True
    def get_session_template(self): return json.dumps(self.to_dict())

    def link_nodes(self, node_name, port_name, target_node_name, target_port_name):
        n1 = self.nodes.get(node_name); n2 = self.nodes.get(target_node_name)
        if n1 and n2:
            p1 = n1.get_port(port_name); p2 = n2.get_port(target_port_name)
            if p1 and p2:
                p2.set_link(p1)
                return True
        return False

    def update(self):
        for node in self.nodes.values(): node.update()

    def to_dict(self):
        objs = DB.dump_all()
        doc = self._document.copy()
        doc["nodes"] = {k: v.oid for k, v in self.nodes.items()}
        return {"session": doc, "objects": objs}

    def save(self, filename):
        """Save session and all registered objects to a JSONL file."""
        with open(filename, 'w') as f:
            # 1. Write session object first (header)
            doc = self._document.copy()
            doc["nodes"] = {k: v.oid for k, v in self.nodes.items()}
            f.write(json.dumps(doc) + "\n")
            
            # 2. Write all other objects currently in DB
            for obj_data in DB.dump_all():
                if obj_data.get("_id") != self.oid:
                    f.write(json.dumps(obj_data) + "\n")

    def append_object(self, obj, filename):
        """Append a single object's state to an existing JSONL session file."""
        with open(filename, 'a') as f:
            if hasattr(obj, "to_dict"):
                data = obj.to_dict()
            elif hasattr(obj, "_document"):
                data = obj._document
            else:
                data = obj
            f.write(json.dumps(data) + "\n")

    @classmethod
    def load(cls, filename):
        """Load session and reconstruct object graph from a JSONL file."""
        with open(filename, 'r') as f:
            lines = f.readlines()
        
        if not lines: return None
        
        # Check if it's legacy monolithic JSON or new JSONL
        try:
            first_line = json.loads(lines[0])
            if not isinstance(first_line, dict): raise ValueError()
            is_jsonl = len(lines) > 1 or "session" not in first_line
        except:
            # Might be old format with one big JSON object
            with open(filename, 'r') as f: data = json.load(f)
            return cls.from_dict(data)

        if not is_jsonl:
            with open(filename, 'r') as f: data = json.load(f)
            return cls.from_dict(data)

        DB.clear()
        id_map = {}
        objects_data = []
        session = None
        
        # Pass 1: Register all raw documents and identify session
        for line in lines:
            if not line.strip(): continue
            data = json.loads(line)
            objects_data.append(data)
            DB.register(data)
            
            if data.get("type") == "session" and session is None:
                session = cls()
                session.set_document(data)
                id_map[session.oid] = session
                DB.register(session)

        # Pass 2: Instantiate Nodes and Ports
        for data in objects_data:
            oid = data.get("_id")
            if not oid or oid in id_map: continue
            
            obj_type = data.get("type")
            name = data.get("name", "")
            
            if obj_type == "port":
                p = Port(name=name, oid=oid)
                p.set_document(data)
                id_map[oid] = p; DB.register(p)
            elif obj_type == "node":
                n = Node(name=name)
                n.oid = oid; n.set_document(data)
                id_map[oid] = n; DB.register(n)

        # Pass 3: Restore Links and Connections
        for data in objects_data:
            obj = id_map.get(data.get("_id"))
            if not obj: continue
            
            if data.get("type") == "port":
                link_oid = data.get("link")
                if link_oid in id_map:
                    obj.set_link(id_map[link_oid])
            elif data.get("type") == "node":
                ports_map = data.get("ports", {})
                for pname, p_oid in ports_map.items():
                    p = id_map.get(p_oid)
                    if p:
                        obj.ports[pname] = p
                        p.set_node(obj)
                obj.fill_input_output_port_lookups()

        # Pass 4: Restore Session's top-level Node map
        if session:
            node_oids = session._document.get("nodes", {})
            for name, n_oid in node_oids.items():
                if n_oid in id_map:
                    session.nodes[name] = id_map[n_oid]

        return session

    @classmethod
    def from_dict(cls, data):
        """Compatibility shim for old monolithic JSON format."""
        DB.clear()
        sess_data = data["session"]; session = cls()
        session.set_document(sess_data); DB.register(session)
        
        for obj_data in data["objects"]:
            DB.register(obj_data)
            
        id_map = {session.oid: session}
        for obj_data in data["objects"]:
            oid = obj_data["_id"]; name = obj_data["name"]; obj_type = obj_data["type"]
            if obj_type == "port":
                p = Port(name=name, oid=oid); p.set_document(obj_data)
                id_map[oid] = p; DB.register(p)
            elif obj_type == "node":
                n = Node(name=name); n.oid = oid; n.set_document(obj_data)
                id_map[oid] = n; DB.register(n)
        
        for obj_data in data["objects"]:
            obj = id_map.get(obj_data["_id"])
            if not obj: continue
            if obj_data["type"] == "port":
                link_oid = obj_data.get("link")
                if link_oid in id_map: obj.set_link(id_map[link_oid])
            elif obj_data["type"] == "node":
                ports_map = obj_data.get("ports", {})
                for pname, p_oid in ports_map.items():
                    p = id_map.get(p_oid)
                    if p: obj.ports[pname] = p; p.set_node(obj)
                obj.fill_input_output_port_lookups()
        
        node_oids = sess_data.get("nodes", {})
        for name, n_oid in node_oids.items():
            if n_oid in id_map: session.nodes[name] = id_map[n_oid]
            
        return session

    def _update_doc_from_data(self, doc):
        doc["nodes"] = {k: v.oid for k, v in self.nodes.items()}
